Prevent user from logging in after 5 consecutive failed login attempts until pw is reset

tw4l opened the pull request:
Fixes #1270 

After 5 consecutive failed logins from the same user, we now prevent the user from logging in even with the correct password until they reset it via their email and present the following message in the frontend:

<img width="422" alt="Screen Shot 2023-10-12 at 6 11 26 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/3481bf53-fd4f-47a2-98f3-da51ced75ba0">

`FailedLogin`s are tracked in a new database collection and set to expire after one hour.

Because `FailedLogin`s are tracked by email, after 5 login attempts for an unregistered email address, the backend will return 429 same as for a registered email address to prevent phishing for registered email addresses. A new test has been added to cover this case.

## To test

Backend tests have been added to the same effect but to test manually with the frontend:

1. Make 5 consecutive failed login attempts for a real user (e.g. `username: admin@example.com, password: invalid`)
2. Verify that response goes from 400 to 429 and message pictured above appears
3. Try to log in with correct password and verify 429 is returned and custom response in the frontend telling you to check your email and reset your password appears
4. Reset password from link in email/logs
5. Verify that you can now log in with that user and the new password
6. Verify that five login attempts for an unregistered email address also returns 429 with the custom response in the frontend telling you to check your email and reset your password appears

SuaYoo reviewed (approved):
Nice one, frontend changes look good!

tw4l commented:
@Shrinks99 I tagged you for a "design" review - would specifically like input on whether or not to make the number of consecutive failed logins configurable. Thanks! :)

Shrinks99 commented:
@tw4l 5 sounds reasonable to me!  Don't think we need to make this one configurable.

ikreymer reviewed (changes requested):
Currently blocked on #1269 and also probably #1290 for full auth refactor. Marking to avoid merging for now.

tw4l commented:
This needs to be rebased/reworked on latest main now that fastapi-users is removed.

tw4l committed:
Lock user after 5 consecutive failed login attempts

After 5 consecutive failed login attempts, send a /forgot-password
email to the user to reset their password. The user is prevented
from logging in even with the correct password until they reset it
and see a "Sorry, too many failed login attempts" message on the
login screen.

After successful authentication, the failed login count is reset
to 0.

tw4l commented:
@ikreymer Rebased on latest main/fastapi-users removal changes and ready for review :)

ikreymer commented on backend/btrixcloud/users.py:
```diff
@@ -434,6 +434,11 @@ async def forgot_password(
             user.email, token, request and request.headers
         )
 
+        # Lock user until password is reset
```
Hm, should this be here? That means if you also manually requested a password reset, but then remembered it later, your account is now locked - do we want to do that? I think should still be able to use account, if you remember password later and choose not to reset it.

tw4l replied:
Removed

ikreymer commented on backend/btrixcloud/auth.py:
```diff
@@ -171,10 +173,52 @@ def get_bearer_response(user: User):
         )
 
+        attempted_user = await user_manager.get_by_email(credentials.username)
+        if attempted_user is None:
```
This theoretically would allow someone to discover if an email is registered in the system or not - if they always get a 400, then its not a valid user, but if they get a 429, then it is.
Not sure of a good way around this, though.

tw4l replied:
Yeah I'm not sure a good way around this either, will think on it.

ikreymer replied:
With the FailedLogin object, we could can now add these regardless of whether its a valid email or not! So using the same invalid email 5 times will get a 429, but won't send a reset if its not a valid user.

tw4l replied:
Added a test to cover this :)

ikreymer commented on backend/btrixcloud/users.py:
```diff
@@ -507,16 +512,37 @@ async def _update_password(self, user: User, new_password: str) -> None:
+    async def inc_failed_logins(self, user: User) -> None:
+        """Inc consecutive failed login attempts for user by 1"""
+        count = await self.get_failed_logins_count(user)
+        await self.set_failed_logins(user, count + 1)
```
Should use mongo `$inc` operator instead.

Maybe the two functions are `inc_failed_logins` which uses inc and `clear_failed_logins` which resets it to 0?

ikreymer commented on backend/btrixcloud/models.py:
```diff
@@ -88,6 +88,14 @@ class User(BaseModel):
+    # If user is logged, they are unable to log in until they reset their
+    # password from the /forgot-password email or /users/me/password-change.
+    # This is to prevent brute force cracking of passwords via /login.
+    locked: Optional[bool] = False
```
Hm, do we actually need a separate `locked` field?

We can just check the `failed_logins > 5` and consider that to be locked.

Completing the reset flow would then set it to 0.

tw4l replied:
It was so that we could reset the failed logins when the password reset email was sent, but if we do that on resetting password instead then yeah, not needed! Will change.

ikreymer commented:
Left a few comments on how to simplify it. But, I think we probably want to use a [TTL Index](https://www.mongodb.com/docs/manual/core/index-ttl/) similar to how we treat invites, and maybe create a FailedLogin object and increment counter on it. After some time, it will expire.
Otherwise, user can be at 4 failed logins and then one failed login locks them out days or months later, which is not ideal.

tw4l committed:
Modify failed_logins methods based on code review feedback

tw4l committed:
Remove User.locked

tw4l committed:
Track login attempts in FailedLogins object with TTL of 1 hour

This allows us to return a 429 after too many attempts on an email
address even if it's not registered to a user, which prevents
attempting to use the /login endpoint to phish for registered emails.

tw4l committed:
Add test for unregistered email address

tw4l commented:
@ikreymer changes made :)

ikreymer committed:
Merge branch 'main' into issue-1270-pw-rate-limiting

ikreymer committed:
minor refactor:
- always check if exceeded max failed threshold first, don't attempt login if already exceeded
- send email only when at exactly 5 attempts, to avoid spamming for every failed login
- uset $setOnInsert to add/increment failed login atomically

ikreymer commented:
Made some more changes:
- Checking failed login threshold first before even attempting to login
- Keep incrementing counter after already failed, only send email the first time to avoid spamming
- Failure now on 6th attempt
- Atomically update in mongodb with a single upsert query.

ikreymer committed:
rename 'created' -> 'attempted', update on each failed login

ikreymer commented:
Further updated to track date via 'created' -> 'attempted', and update after every attempt, requiring an hour of no failed logins.
Currently not permanently locking account, just disabling logins for an hour. I think that's probably a good setup for now, can revisit permanent locking if need be.
